backend/src/load.rs

```diff
@@ -0,0 +1,404 @@
+use std::collections::{HashMap, HashSet};
+use std::cmp::{Ordering, max};
+use std::io::Read;
+
+use chrono::{Duration, NaiveDateTime};
+use json::{self, JsonValue};
+
+use util::start_idx;
+use errors::*;
+
+const WEEKS_IN_SUMMARY: usize = 12;
+
+/// Contains the data loaded from the file system.
+/// Useful to avoid passing around each individual field, instead passing the
+/// entire struct.
+pub struct ServerData {
+    pub summary_rustc: Summary,
+    pub summary_benchmarks: Summary,
+
+    /// A set containing all crate names.
+    pub crate_list: HashSet<String>,
+
+    /// A set containing all phase names, across all crates.
+    pub phase_list: HashSet<String>,
+
+    /// TODO: Better docs. It's unknown how this is used on the client.  A set
+    /// of test names. Test names are found from the file path, everything
+    /// before the `--` is included.
+    pub benchmarks: HashSet<String>,
+
+    /// The last date that was seen while loading files. The DateTime variant is
+    /// used here since the date may or may not contain a time. Since the
+    /// timezone is not important, it isn't stored, hence the Naive variant.
+    pub last_date: NaiveDateTime,
+
+    /// Private due to access being exposed through by_kind method.
+    /// Care must be taken to sort these after modifying them.
+    data_rustc: Vec<Data>,
+    data_benchmarks: Vec<Data>,
+}
+
+/// Allows storing ServerData in Iron's persistent data structure.
+impl ::iron::typemap::Key for ServerData { type Value = ServerData; }
+
+impl ServerData {
+    /// Initialize ServerData from the file system.
+    pub fn from_fs() -> Result<ServerData> {
+        use std::path::Path;
+        use std::fs::{self, File};
+
+        // TODO: Read this at runtime, not hardcoded.
+        let repo_loc = Path::new("../data");
+
+        let mut last_date = None;
+        let mut crate_list = HashSet::new();
+        let mut phase_list = HashSet::new();
+        let mut no_times = 0;
+        let mut c_bootstraps = 0;
```

Comment:
This is the same near as I can tell, but `c_benchmarks` isn't (unless we don't care about tracking `c_benchmarks_add` separately). I'll remote `c_bootstraps`, not sure if the tracking on c_benchmarks_add is important.